File: himsha-sdk-dart/lib/src/connection.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'pubkey.dart';
import 'transaction.dart';

class HimshaRpcException implements Exception {
  final int code;
  final String message;
  HimshaRpcException(this.code, this.message);
  @override
  String toString() => 'HimshaRpcException($code): $message';
}

class AccountInfo {
  final String key;
  final BigInt lamports;
  final String dataHex;
  final String owner;
  final bool executable;

  const AccountInfo({
    required this.key,
    required this.lamports,
    required this.dataHex,
    required this.owner,
    required this.executable,
  });

  factory AccountInfo.fromJson(Map<String, dynamic> j) => AccountInfo(
        key:        j['key'] as String,
        lamports:   BigInt.parse(j['lamports'].toString()),
        dataHex:    j['data'] as String,
        owner:      j['owner'] as String,
        executable: j['executable'] as bool,
      );
}

class UtxoInfo {
  final String txid;
  final int vout;
  final BigInt value;
  final String scriptPubkey;
  final int confirmations;

  const UtxoInfo({
    required this.txid,
    required this.vout,
    required this.value,
    required this.scriptPubkey,
    required this.confirmations,
  });

  factory UtxoInfo.fromJson(Map<String, dynamic> j) {
    final meta = j['meta'] as Map<String, dynamic>;
    return UtxoInfo(
      txid:          meta['txid'] as String,
      vout:          meta['vout'] as int,
      value:         BigInt.parse(j['value'].toString()),
      scriptPubkey:  j['scriptPubkey'] as String,
      confirmations: j['confirmations'] as int,
    );
  }
}

/// HIMSHA Network JSON-RPC client for Dart.
///
/// ```dart
/// final conn = HimshaConnection('http://localhost:9100');
/// final ready = await conn.isNodeReady();
/// ```
class HimshaConnection {
  final String endpoint;
  int _nextId = 1;
  final http.Client _client;

  HimshaConnection(this.endpoint, {http.Client? client})
      : _client = client ?? http.Client();

  Future<T> _call<T>(String method, [List<dynamic> params = const []]) async {
    final id = _nextId++;
    final body = jsonEncode({
      'jsonrpc': '2.0',
      'id': id,
      'method': method,
      'params': params,
    });

    final response = await _client.post(
      Uri.parse(endpoint),
      headers: {'Content-Type': 'application/json'},
      body: body,
    );

    if (response.statusCode != 200) {
      throw HimshaRpcException(response.statusCode, 'HTTP ${response.statusCode}');
    }

    final json = jsonDecode(response.body) as Map<String, dynamic>;
    if (json.containsKey('error')) {
      final err = json['error'] as Map<String, dynamic>;
      throw HimshaRpcException(err['code'] as int, err['message'] as String);
    }
    return json['result'] as T;
  }

  // ---- Node queries ----

  Future<bool> isNodeReady() => _call<bool>('himsha_isNodeReady');

  Future<BigInt> getSlot() async {
    final slot = await _call<dynamic>('himsha_getSlot');
    return BigInt.parse(slot.toString());
  }

  Future<Map<String, dynamic>?> getBlock(BigInt slot) =>
      _call<Map<String, dynamic>?>('himsha_getBlock', [slot.toString()]);

  Future<List<String>> listPrograms() =>
      _call<List<dynamic>>('himsha_listPrograms').then((l) => l.cast<String>());

  // ---- Account queries ----

  Future<AccountInfo?> getAccountInfo(HimshaPublicKey pubkey) async {
    final raw = await _call<Map<String, dynamic>?>('himsha_getAccountInfo', [pubkey.toBase58()]);
    return raw == null ? null : AccountInfo.fromJson(raw);
  }

  Future<List<AccountInfo>> getProgramAccounts(HimshaPublicKey programId) async {
    final raw = await _call<List<dynamic>>('himsha_getProgramAccounts', [programId.toBase58()]);
    return raw
        .cast<Map<String, dynamic>>()
        .map(AccountInfo.fromJson)
        .toList();
  }

  Future<bool> accountExists(HimshaPublicKey pubkey) async =>
      (await getAccountInfo(pubkey)) != null;

  // ---- Bitcoin ----

  Future<UtxoInfo?> getUtxo(String txid, int vout) async {
    final raw = await _call<Map<String, dynamic>?>('himsha_getUtxo', [txid, vout]);
    return raw == null ? null : UtxoInfo.fromJson(raw);
  }

  // ---- Transactions ----

  Future<String> sendTransaction(HimshaTransaction tx) =>
      _call<String>('himsha_sendTransaction', [tx.toJson()]);

  Future<String> deployProgram(String elfHex, String imageIdHex) =>
      _call<String>('himsha_deployProgram', [elfHex, imageIdHex]);

  /// Execution status of a submitted transaction (snake_case `{status, slot,
  /// error}`), or `null` if the node has never seen the id. Since execution
  /// happens at block production (not at submit time), this is how a client
  /// learns the authoritative outcome.
  Future<Map<String, dynamic>?> getSignatureStatus(String txId) =>
      _call<Map<String, dynamic>?>('himsha_getSignatureStatus', [txId]);

  /// Merkle inclusion proof (snake_case `{state_root, leaf, index, siblings,
  /// anchored_slot, anchored_state_root, anchored_btc_txid}`) that `pubkey`'s
  /// account is committed in the current state root. Returns `null` if the
  /// account does not exist.
  Future<Map<String, dynamic>?> getStateProof(HimshaPublicKey pubkey) =>
      _call<Map<String, dynamic>?>('himsha_getStateProof', [pubkey.toBase58()]);

  /// Threshold-custody status (snake_case `{threshold, total, group_key,
  /// address}`): the M-of-N config, the committee group key, and the Taproot
  /// address to fund. Returns `null` in single-hot-wallet mode.
  Future<Map<String, dynamic>?> getCustodyInfo() =>
      _call<Map<String, dynamic>?>('himsha_getCustodyInfo');

  /// Poll until a transaction is executed. Resolves with the slot once it
  /// `succeeded`; **throws with the failure reason** if it `failed` (no more
  /// silent timeouts on a rejected tx); keeps waiting while `pending`; times
  /// out otherwise.
  Future<BigInt> confirmTransaction(String txId,
      {Duration timeout = const Duration(seconds: 30)}) async {
    final deadline = DateTime.now().add(timeout);
    while (DateTime.now().isBefore(deadline)) {
      final st = await getSignatureStatus(txId);
      final status = st?['status'] as String?;
      if (status == 'succeeded') {
        final slot = st?['slot'];
        return slot == null ? BigInt.zero : BigInt.parse(slot.toString());
      }
      if (status == 'failed') {
        final slot = st?['slot'];
        final at = slot != null ? ' at slot $slot' : '';
        final reason = (st?['error'] as String?) ?? 'unknown error';
        throw HimshaTransactionFailedException(
            'Transaction $txId failed$at: $reason');
      }
      await Future.delayed(const Duration(milliseconds: 500));
    }
    throw TimeoutException(
        'Transaction $txId not confirmed within $timeout', timeout);
  }

  void close() => _client.close();
}

class HimshaTransactionFailedException implements Exception {
  final String message;
  HimshaTransactionFailedException(this.message);
  @override
  String toString() => 'HimshaTransactionFailedException: $message';
}

class TimeoutException implements Exception {
  final String message;
  final Duration? timeout;
  TimeoutException(this.message, [this.timeout]);
  @override
  String toString() => 'TimeoutException: $message';
}

File: himsha-sdk-python/himsha_sdk/connection.py
"""Async JSON-RPC connection to a HIMSHA Network node."""

from __future__ import annotations

import asyncio
from dataclasses import dataclass
from typing import Any, TypeVar

import httpx

from .pubkey import HimshaPublicKey
from .transaction import HimshaTransaction

T = TypeVar("T")


class HimshaRpcError(Exception):
    def __init__(self, code: int, message: str) -> None:
        super().__init__(f"RPC error {code}: {message}")
        self.code = code


class HimshaTransactionFailed(Exception):
    """Raised by :meth:`HimshaConnection.confirm_transaction` when a transaction
    executed but its on-chain status is ``failed``."""


@dataclass
class AccountInfo:
    key:        str
    lamports:   int
    data_hex:   str
    owner:      str
    executable: bool

    @property
    def data(self) -> bytes:
        return bytes.fromhex(self.data_hex)

    @classmethod
    def from_json(cls, j: dict) -> "AccountInfo":
        return cls(
            key=j["key"],
            lamports=int(j["lamports"]),
            data_hex=j["data"],
            owner=j["owner"],
            executable=j["executable"],
        )


@dataclass
class UtxoInfo:
    txid:          str
    vout:          int
    value:         int   # satoshis
    script_pubkey: str
    confirmations: int

    @classmethod
    def from_json(cls, j: dict) -> "UtxoInfo":
        meta = j["meta"]
        return cls(
            txid=meta["txid"],
            vout=meta["vout"],
            value=int(j["value"]),
            script_pubkey=j["scriptPubkey"],
            confirmations=j["confirmations"],
        )


class HimshaConnection:
    """
    Async JSON-RPC client for HIMSHA Network.

    Usage::

        async with HimshaConnection("http://localhost:9100") as conn:
            ready = await conn.is_node_ready()
            slot  = await conn.get_slot()
    """

    def __init__(self, endpoint: str, timeout: float = 30.0) -> None:
        self.endpoint = endpoint
        self._client  = httpx.AsyncClient(timeout=timeout)
        self._next_id = 1

    async def __aenter__(self) -> "HimshaConnection":
        return self

    async def __aexit__(self, *_: Any) -> None:
        await self.close()

    # ---- low-level ----

    async def _call(self, method: str, params: list[Any] | None = None) -> Any:
        req_id = self._next_id
        self._next_id += 1
        body = {"jsonrpc": "2.0", "id": req_id, "method": method, "params": params or []}
        response = await self._client.post(
            self.endpoint,
            json=body,
            headers={"Content-Type": "application/json"},
        )
        response.raise_for_status()
        data = response.json()
        if "error" in data:
            err = data["error"]
            raise HimshaRpcError(err["code"], err["message"])
        return data["result"]

    # ---- node ----

    async def is_node_ready(self) -> bool:
        return await self._call("himsha_isNodeReady")

    async def get_slot(self) -> int:
        slot = await self._call("himsha_getSlot")
        return int(slot)

    async def get_block(self, slot: int) -> dict | None:
        return await self._call("himsha_getBlock", [str(slot)])

    async def list_programs(self) -> list[str]:
        return await self._call("himsha_listPrograms")

    # ---- accounts ----

    async def get_account_info(self, pubkey: HimshaPublicKey | str) -> AccountInfo | None:
        key = str(pubkey)
        raw = await self._call("himsha_getAccountInfo", [key])
        return AccountInfo.from_json(raw) if raw else None

    async def get_program_accounts(self, program_id: HimshaPublicKey | str) -> list[AccountInfo]:
        pid = str(program_id)
        raw_list = await self._call("himsha_getProgramAccounts", [pid])
        return [AccountInfo.from_json(r) for r in (raw_list or [])]

    async def account_exists(self, pubkey: HimshaPublicKey | str) -> bool:
        return (await self.get_account_info(pubkey)) is not None

    # ---- bitcoin ----

    async def get_utxo(self, txid: str, vout: int) -> UtxoInfo | None:
        raw = await self._call("himsha_getUtxo", [txid, vout])
        return UtxoInfo.from_json(raw) if raw else None

    # ---- transactions ----

    async def send_transaction(self, tx: HimshaTransaction) -> str:
        """Submit a signed transaction. Returns the transaction ID (hex)."""
        return await self._call("himsha_sendTransaction", [tx.to_json()])

    async def deploy_program(self, elf_hex: str, image_id_hex: str) -> str:
        """Deploy a compiled ELF. Returns the program's public key (base58)."""
        return await self._call("himsha_deployProgram", [elf_hex, image_id_hex])

    async def get_signature_status(self, txid: str) -> dict | None:
        """
        Execution status of a submitted transaction, or ``None`` if the node has
        never seen the id. Since execution happens at block production (not at
        submit time), this is how a client learns the authoritative outcome.

        Returns a dict ``{status, slot, error}`` where ``status`` is one of
        ``"pending"``, ``"succeeded"``, ``"failed"``.
        """
        return await self._call("himsha_getSignatureStatus", [txid])

    async def get_state_proof(self, pubkey: HimshaPublicKey | str) -> dict | None:
        """
        State-root inclusion proof for an account, or ``None`` if not found.

        Returns a dict ``{state_root, leaf, index, siblings, anchored_slot,
        anchored_state_root, anchored_btc_txid}``. Verify it with
        ``himsha_sdk.state_proof.verify_state_proof`` /
        ``verify_account_in_state``.
        """
        key = str(pubkey)
        return await self._call("himsha_getStateProof", [key])

    async def get_custody_info(self) -> dict | None:
        """
        Threshold-custody settlement configuration, or ``None`` if custody is not
        configured. Returns a dict ``{threshold, total, group_key, address}``.
        """
        return await self._call("himsha_getCustodyInfo")

    async def confirm_transaction(
        self, tx_id: str, timeout_secs: float = 30.0
    ) -> int:
        """
        Poll until a transaction is executed. Returns the slot once it
        ``succeeded``; **raises** with the failure reason if it ``failed`` (no
        silent timeouts on a rejected tx); keeps waiting while ``pending``;
        raises :class:`TimeoutError` if it never settles in time.
        """
        deadline = asyncio.get_event_loop().time() + timeout_secs
        while asyncio.get_event_loop().time() < deadline:
            st = await self.get_signature_status(tx_id)
            if st is not None:
                status = st.get("status")
                if status == "succeeded":
                    return int(st.get("slot") or 0)
                if status == "failed":
                    slot = st.get("slot")
                    at = f" at slot {slot}" if slot is not None else ""
                    reason = st.get("error") or "unknown error"
                    raise HimshaTransactionFailed(
                        f"Transaction {tx_id} failed{at}: {reason}"
                    )
            await asyncio.sleep(0.5)
        raise TimeoutError(f"Transaction {tx_id} not confirmed within {timeout_secs}s")

    async def close(self) -> None:
        await self._client.aclose()
